docs/architecture/psa-migration/psa-legacy-bridges.md

```diff
@@ -0,0 +1,321 @@
+                               const psa_key_attributes_t *attributes,
+                               mbedtls_svc_key_id_t *key_id);
+```
+
+* `mbedtls_pk_get_psa_attributes` does not change the id/lifetime fields of the attributes (which indicate a volatile key by default).
+* `mbedtls_pk_get_psa_attributes` sets the type and size based on what's in the pk context.
+    * The key type is a key pair if the context contains a private key, and a public key if the context only contains a public key.
+* `mbedtls_pk_get_psa_attributes` sets all the potentially applicable usage flags: `EXPORT`, `COPY`; `VERIFY_HASH | VERIFY_MESSAGE` or `ENCRYPT` as applicable for both public keys and key pairs; `SIGN` or `DECRYPT` as applicable for a key pair.
+* [OPEN] What is the default algorithm for `mbedtls_pk_get_psa_attributes`? Suggestion: assume signature by default. For RSA, either `PSA_RSA_PKCS1_V15_SIGN(PSA_ALG_ANY_HASH)` or `PSA_ALG_RSA_PSS(hash_alg)` depending on the RSA context's padding mode. For ECC, `PSA_ALG_DETERMINISTIC_ECDSA` if `MBEDTLS_ECDSA_DETERMINISTIC` is enabled and `PSA_ALG_ECDSA` otherwise.
+* [OPEN] Or does `mbedtls_pk_get_psa_attributes` need an extra argument indicating how to treat RSA and ECC keys?
+* `mbedtls_pk_import_into_psa` checks that the type field in the attributes is consistent with the content of the `mbedtls_pk_context` object (RSA/ECC, and availability of the private key).
+    * The key type can be a public key even if the private key is available.
+* `mbedtls_pk_import_into_psa` does not need to check the bit-size in the attributes: `psa_import_key` will do enough checks.
+* `mbedtls_pk_import_into_psa` does not check that the policy in the attributes is sensible. That's on the user.
+
+#### API to copy a PSA key to a PK context
+
+Based on the [gap analysis](#using-a-psa-key-as-a-pk-context):
+
+[ACTION] Implement `mbedtls_pk_copy_from_psa` as described below.
+
+```
+int mbedtls_pk_copy_from_psa(mbedtls_svc_key_id_t key_id,
+                             mbedtls_pk_context *pk);
+```
+
+* `pk` must be initialized, but not set up.
+* It is an error if the key is neither a key pair nor a public key.
+* It is an error if the key is not exportable.
+* The resulting pk object has a transparent type, not `MBEDTLS_PK_OPAQUE`.
+* Once this function returns, the pk object is completely independent of the PSA key.
+* Calling `mbedtls_pk_sign`, `mbedtls_pk_verify`, `mbedtls_pk_encrypt`, `mbedtls_pk_decrypt` on the resulting pk context will perform an algorithm that is compatible with the PSA key's primary algorithm policy (`psa_get_key_algorithm`), but with no restriction on the hash (as if the policy had `PSA_ALG_ANY_HASH` instead of a specific hash, and with `PSA_ALG_RSA_PKCS1V15_SIGN_RAW` merged with `PSA_ALG_RSA_PKCS1V15_SIGN(hash_alg)`). For ECDSA, the choice of deterministic vs randomized will be based on the compile-time setting `MBEDTLS_ECDSA_DETERMINISTIC`, like `mbedtls_pk_sign` today.
```

Comment:
This isn't about policies: it doesn't prevent a bypass with `pk_{sign,verify}_ext`. This is about what happens when you call a pk function that doesn't specify an algorithm. If you have code that's currently using a PK object, and you now want to switch to PSA for constructing the object (e.g. to benefit from a PSA driver on the new generation of your hardware) but the key is still used by an API that consumes pk object, then the meaning of `mbedtls_pk_{sign,verify,encrypt,decrypt}` must not change.

Reply:
Ah, right, thanks for clarifying, I had missed the point indeed!

Reply:
IMO the least surprising thing to do would be for keys created with `pk_copy_from_psa()` to behave the same as keys created with `pk_parse_xxx()`. That is, `pk_copy_from_psa()` should be fully equivalent to:
1. set up temporary context with `pk_setup_opaque()`;
2. use `pk_write_xxx()` on it then `free()` it;
3. populate the target context with `pk_parse_xxx()` on last step's output.

As a user, I wouldn't like the result of the above sequence and that of `pk_copy_from_psa()` to differ in an observable way.

Reply:
I'm not sure I agree that `pk_copy_from_psa` should be equivalent to write+parse. There's algorithm-usage information in the PSA object that might not be expressible in the export format, but is partly expressible in a pk object.

Reply:
Hmm, I see your point. OTOH, "partly" is the key word here, so I'd like to explore what we could do.

For RSA keys:
- We could call `mbedtls_rsa_set_padding()` based on the alg. However how do we handle (1) `PSA_ALG_ANY_HASH` (likely means passing `MD_NONE` to `set_padding()`?) and (2) the distinction between `SIGN` and `SIGN_RAW`?
  - For OAEP, `ANY_HASH` is not an option so that's great.
  - For PSS generation, `mbedtls_rsa_rsassa_pss_sign_ext()` and `mbedtls_rsa_rsassa_pss_sign()` return an error when `hash_id` is `NONE` but we've recently introduced `mbedtls_rsa_rsassa_pss_sign_no_mode_check()` so we could use that.
  - For PSS verification, `NONE` is accepted so that' OK too.
- I don't think we can really do anything based on the policy though.

For ECC keys:
- Based on the alg (and policy), we could select either `ECDSA` or `ECKEY_DH` as the type.
- But we can't really do anything about the distinction between deterministic and random though.

So, on one hand I agree that if you have an RSA key with an `ALG_PSS` policy, and you call `pk_copy_from_psa()` on it then `pk_sign()` with the result, it's a pretty reasonable expectation that you should get a PSS signature.

On the other hand, I'm a bit tired of writing documentation (and tests) about things that work partly, and where it's not trivial even for us to get a good intuition of what's happening and we keep having to check the documentation because it's too complex for us to remember. So, what I like about the "equivalent to write+parse" option is it's simple to describe and (for me at least) know how it works. But I agree it can also be surprising and counter to some people's reasonable expectations.

Reply:
>  I'm a bit tired of writing documentation (and tests) about things that work partly

I agree. I can't finalize the specification here without having written a tested implementation, because I want to make sure that I'm not requiring something impossible. But I think I'm on a reasonable track.

> PSA_ALG_ANY_HASH (likely means passing MD_NONE to set_padding()?)

Yes, I think so. The hash is overridden when you call `pk_sign` anyway.

> the distinction between SIGN and SIGN_RAW?

I don't think the legacy API makes a distinction in the key data: that's always decided when calling `pk_sign`.

> But we can't really do anything about the distinction between deterministic and random though.

Right. I think that's an acceptable deviation (deciding at runtime in PSA, at compile-time in legacy) since there's no functional difference.